File: frontend/src/types/sqlFlow.ts
import type { DbValue } from '@/types/database';

export interface SqlFlowNode {
  id: string;
  label: string;
  isRoot: boolean;
  alias?: string;
  columns?: string[];
  rows?: Record<string, DbValue>[];
}

export interface SqlFlowEdge {
  id: string;
  source: string;
  target: string;
  label: string;
  joinType: string;
}

export interface SqlFlowGraph {
  nodes: SqlFlowNode[];
  edges: SqlFlowEdge[];
}

export interface ParseSqlFlowParams {
  query: string;
  dialect?: string;
}

export interface GetSqlFlowDataParams {
  query: string;
  connectionId?: string;
  schema?: string;
  dialect?: string;
}

File: src-tauri/src/models/sql_flow.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SqlFlowNode {
    pub id: String,
    pub label: String,
    pub is_root: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alias: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub columns: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rows: Option<Vec<serde_json::Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SqlFlowEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    pub label: String,
    pub join_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct SqlFlowGraph {
    pub nodes: Vec<SqlFlowNode>,
    pub edges: Vec<SqlFlowEdge>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParseSqlFlowRequest {
    pub query: String,
    #[serde(default)]
    pub dialect: Option<String>,
}
